runtime/parachains/src/runtime_api_impl/vstaging.rs

```diff
@@ -26,6 +29,50 @@ pub fn get_session_disputes<T: disputes::Config>(
 	<disputes::Pallet<T>>::disputes()
 }
 
-// TODO [now]: implicit `validity_constraints`. Ensure that `min_relay_parent`
-// never goes lower than the point at which asynchronous backing was enabled.
-// Also, never cross session boundaries.
+/// Implementation for `StagingValidityConstraints` function from the runtime API
+pub fn validity_constraints<T: initializer::Config>(
+	para_id: ParaId,
+) -> Option<Constraints<T::BlockNumber>> {
+	// Async backing is only expected to be enabled with a tracker capacity of 1.
+	// Subsequent configuration update gets applied on new session, which always
+	// clears the buffer.
+	//
+	// Thus, minimum relay parent is ensured to have asynchronous backing enabled.
+	let min_relay_parent_number =
+		<shared::Pallet<T>>::allowed_relay_parents().earliest_block_number()?;
```

Comment:
Depending on the behavior of runtime APIs, which I'm a bit rusty on, this may be `None` on the first block of the session.

I think it used to be that runtime APIs invoked `on_initialize` but I don't believe they do anymore.

The `shared` module clears the buffer, as the comment above states, here: https://github.com/paritytech/polkadot/blob/21e19328bd8d3130acf50394de2ec4c19cdcaa67/runtime/parachains/src/shared.rs#L180

This happens at the end of the block of the last session. Therefore, since `on_initialize` hasn't been called, the buffer will be empty. This will lead to `None` return for this function, which might cause spurious logs here: 

https://github.com/paritytech/polkadot/blob/21e19328bd8d3130acf50394de2ec4c19cdcaa67/node/core/prospective-parachains/src/lib.rs#L215-L226

Reply:
Since the `availability_cores` Runtime API implementation invokes `Scheduler::schedule`, it seems to me that we will have 'scheduled' cores.

I don't think we want the `fn validity_constraints` to return `None` on the first block of new sessions, so we should probably fall back to returning a `min_relay_parent_number = now` (given that this is called against the post-state of the parent)

Reply:
Actually, it seems weird altogether that we only invoke `AllowedRelayParents::update` within the `paras_inherent`, and it seems to me that the logic of updating should emulate the update logic to the extent possible

Reply:
Can you elaborate on 
> the logic of updating should emulate the update logic to the extent possible

?

Reply:
Basically the beginning of the runtime API here needs to do `AllowedRelayParents::update`, both to add the new thing into the buffer and also to remove anything outgoing. Except it can't exactly do `update` because it doesn't have anything other than the parent block number (`= system::block_number()`). So it needs to invoke some new method on `AllowedRelayParents` which is like `hypothetical_earliest_relay_parent(parent_block_number, ALLOWED_RELAY_PARENT_LOOKBACK) -> BlockNumber`